supabase/migrations/14_gateway_auth.sql

```diff
@@ -0,0 +1,71 @@
+-- Keys used to sign/verify gateway auth tokens.
+create table internal.gateway_auth_keys (
+  like internal._model including all,
+  -- Key used to sign JWTs
+  secret_key text
+);
+
+insert into internal.gateway_auth_keys (secret_key, detail) values (
+  'supersecret', 'Used for development only. This value will be changed manually when deployed to production.'
+);
+
+
+
+-- Addresses of deployed data plane gateways. As we deploy into multiple
+-- AZs/Regions, we can direct a caller to the appropriate Gateway for accessing
+-- data in a region-aware way.
+create table internal.gateway_endpoints (
+  like internal._model including all,
+  name text,
+  url text
+);
+
+insert into internal.gateway_endpoints (name, url, detail) values (
+  'local', 'http://localhost:28318/', 'Used for development only. This value will be changed manually when deployed to production.'
+);
+
+
+
+-- Returns the most appropriate gateway url. For now, there should only be one.
+create function internal.gateway_endpoint_url()
+returns text as $$
+
+  select url
+  from internal.gateway_endpoints
+  limit 1
+
+$$ language sql stable security definer;
+
+
+
+-- Grabs the secret signing key and signs the object.
+create function internal.sign_jwt(obj json)
+returns text as $$
+
+  select sign(obj, secret_key::text)
+  from internal.gateway_auth_keys
+  limit 1
+
+$$ language sql stable security definer;
+
+
+
+create function gateway_auth_token(variadic prefixes text[])
+returns table (token text, gateway_url text) as $$
+
+  select internal.sign_jwt(
```

Comment:
no action required immediately, jotting down thoughts for later / conversation:

- this currently requires that `prefixes` exactly match a granted role. If you passed in a prefix which was _under_ a granted role, it would silently be omitted from the JWT. For consistency with auth checks elsewhere, I'd recommend a select over `unnest(prefixes)` with a [`auth_catalog(prefix, 'read')` predicate.](https://github.com/estuary/animated-carnival/blob/main/supabase/migrations/04_auth.sql#L154). That function is intended to be the primary place we're verifying access authorization.
- we should probably raise an exception if the user requests a prefix to which they're not authorized.

Reply:
Good catch. I'm not sure where in my refactoring I lost it, but I'd had the `starts_with` prefix matching in there originally.

I'm not actually clear how I'd raise an exception here. Right now it simply generates a safe result set that only has valid, authorized prefixes.

```console
curl -s -X POST "http://localhost:5431/rest/v1/rpc/gateway_auth_token" \
      -H "Content-Type: application/json" \
      -H "apikey: $ANON_KEY" \
      -H "Authorization: Bearer $SUPABASE_KEY" \
      -d '{ "prefixes": ["acmeCo/foo", "ops/bobCo/", "ops/", "accessDenied/", "acmeCo/bar/baz"] }' | jq '.[0].token' -r | jwt decode -

Token header
------------
{
  "typ": "JWT",
  "alg": "HS256"
}

Token claims
------------
{
  "exp": 1652977096,
  "iat": 1652973496,
  "operation": "read",
  "prefixes": [
    "acmeCo/bar/baz",
    "acmeCo/foo",
    "ops/bobCo/"
  ],
  "sub": "22222222-2222-2222-2222-222222222222"
}```

Reply:
(Again, this can wait. The immediate goal is to get this in production so UI development can begin. I'm just giving feedback on improvements from here).